File: exercise/C#/day24/src/Delivery/Domain/StockUnit.cs
using Delivery.Domain.Core;
using LanguageExt;
using static Delivery.Domain.Core.Error;

namespace Delivery.Domain
{
    public readonly struct StockUnit(int value)
    {
        public int Value { get; } = value;

        public static Either<Error, StockUnit> From(int stock)
            => stock >= 0
                ? new StockUnit(stock)
                : AnError("A stock unit can not be negative");

        public bool IsSupplied() => Value > 0;
        public StockUnit Decrease() => new(Value - 1);
    }
}

File: exercise/php/day24/src/Delivery/Domain/StockUnit.php
<?php

namespace Delivery\Domain;

class StockUnit
{
    private int $value;

    private function __construct(int $value)
    {
        $this->value = $value;
    }

    /**
     * @throws \Exception
     */
    public static function from(int $stock): StockUnit
    {
        return $stock >= 0
            ? new self($stock)
            : throw new \Exception('A stock unit cannot be negative');
    }

    public function isSupplied(): bool
    {
        return $this->value > 0;
    }

    public function decrease(): self
    {
        return new self($this->value - 1);
    }

    public function getValue(): int
    {
        return $this->value;
    }
}
